File: 036_PathSumIII.cpp
// Topic: Binary Tree / Prefix sum + DFS
// Path Sum III: Count number of paths that sum to target (paths can start/end anywhere downward).
// Time Complexity: O(n)
// Space Complexity: O(n)

#include <bits/stdc++.h>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int dfs(TreeNode* node, long long curr, int target, unordered_map<long long, int>& pref)
{
    if (!node)
        return 0;
    curr += node->val;
    int cnt = 0;
    if (pref.count(curr - target))
        cnt += pref[curr - target];
    ++pref[curr];
    cnt += dfs(node->left, curr, target, pref);
    cnt += dfs(node->right, curr, target, pref);
    --pref[curr];
    return cnt;
}

int pathSum(TreeNode* root, int targetSum)
{
    unordered_map<long long, int> pref;
    pref[0] = 1;
    return dfs(root, 0, targetSum, pref);
}

int main()
{
    TreeNode* root = new TreeNode(10);
    root->left = new TreeNode(5);
    root->right = new TreeNode(-3);
    root->left->left = new TreeNode(3);
    root->left->right = new TreeNode(2);
    root->left->left->left = new TreeNode(3);
    root->left->left->right = new TreeNode(-2);
    root->left->right->right = new TreeNode(1);
    root->right->right = new TreeNode(11);
    cout << pathSum(root, 8) << "\n";  // 3
    TreeNode* r2 = new TreeNode(1);
    r2->left = new TreeNode(-2);
    r2->right = new TreeNode(-3);
    cout << pathSum(r2, -1) << "\n";
    return 0;
}

File: 036_PathSumIII.py
"""
Topic: Tree / Prefix Sum + DFS
Path Sum III: count paths summing to target using prefix-sum hashmap
Time Complexity: O(n)
Space Complexity: O(n)
"""

from typing import Optional
from collections import defaultdict

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def path_sum(root: Optional[TreeNode], target: int) -> int:
    pref = defaultdict(int)
    pref[0] = 1
    def dfs(node, cur):
        if not node: return 0
        cur += node.val
        cnt = pref[cur - target]
        pref[cur] += 1
        cnt += dfs(node.left, cur) + dfs(node.right, cur)
        pref[cur] -= 1
        return cnt
    return dfs(root, 0)

if __name__ == '__main__':
    # [10,5,-3,3,2,null,11,3,-2,null,1]
    root = TreeNode(10, TreeNode(5, TreeNode(3, TreeNode(3), TreeNode(-2)), TreeNode(2, None, TreeNode(1))), TreeNode(-3, None, TreeNode(11)))
    print(path_sum(root, 8))  # 3
    print(path_sum(None, 0))  # 0
